python/sdk/docs/tutorial.rst

```diff
@@ -0,0 +1,666 @@
+        #   - the first item will be an integer that represents the number of samples on which the participant trained
+        #   - the second item represents the model trained by the participant
+        (number_of_samples, weights) = training_result
+
+        # The writer holds the buffer into which we'll write the serialized data
+        writer = BytesIO()
+
+        # The coordinator expects the number of samples to be encoded on 4 bytes, in BigEndian
+        writer.write(number_of_samples.to_bytes(4, byteorder="big"))
+
+        # Append the numpy array
+        np.save(writer, weights, allow_pickle=False)
+
+        # Return the entire buffer
+        return writer.getbuffer()[:]
+
+Training
+^^^^^^^^
+
+We can now focus on the most interesting method: the one where
+training happens. In our case, we'll just generate partially random
+data as explained in `the "Goals" section <goals>`_.
+
+We want to generate an array of 5 identical float numbers, between 0
+and some upper bound from the latest global model:
+
+.. code-block:: Python
+
+   import random
+
+   # ...
+
+   def train_round(self, global_weights: np.ndarray) -> Tuple[int, np.ndarray]:
+       # Get the upper bound from the global model:
+       self.upper_bound = global_weights[0]
+
+       # Pick a random value
+       value = random.uniform(0.0, self.upper_bound)
+
+       # Create the model to send to the coordinator
+       local_weights = np.repeat(value, 5)
+
+       # The coordinator also expects the number of samples the
+       # participant trained on, but we're not actually doing any
+       # training, so let's hardcode this to 1
+       number_of_samples = 1
+
+       return (number_of_samples, local_weights)
+
+Starting the participant
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, our ``main()`` function doesn't do anything apart from
+parsing the CLI arguments. Let's instantiate our participant, and
+start it with ``xain_sdk.run_participant``. We also set up some logger
+with ``xain_sdk.configure_logging``:
+
+.. code-block:: Python
+
+    def main():
+        logging.basicConfig(
+            format="%(asctime)s.%(msecs)03d %(levelname)-8s %(message)s",
+            level=logging.DEBUG,
+            datefmt="%Y-%m-%d %H:%M:%S",
+        )
+
+        parser = argparse.ArgumentParser(description="run a participant")
+        parser.add_argument(
+            "--upper-bound",
+            required=True,
+            type=float,
+            help="Initial upper bound for picking a random float to send to the coordinator",
+        )
+        args = parser.parse_args()
+
+        # Instantiate a participant
+        participant = Participant(args.upper_bound)
+
+        # Set up some logger so that we can see the requests being made to the coordinator
+        xain_sdk.configure_logging(log_http_requests=True)
+
+        # Start the participant
+        coordinator_url = "http://localhost:8081"
+        xain_sdk.run_participant(participant, coordinator_url)
```

Comment:
In the code blocks it would be nice to have a new line at the end to make it easier to just copy and past the code.

Reply:
I added them, but even when adding some indentation, they seem to be stripped unfortunately.